catwalk/models/rank_classification.py

```diff
@@ -202,46 +216,198 @@ def _run_loglikelihood(
+        else:
+            return self._reorder_by_longest(tokenized_contexts, tokenized_continuations)
+    
+    def _reorder_by_prefix(self, tokenized_contexts: BatchEncoding, tokenized_continuations: BatchEncoding) -> Sequence[int]:
+        self._reset_cache_variables()
+        combined_ids = [context + continuation for context, continuation in zip(tokenized_contexts['input_ids'], tokenized_continuations['input_ids'])]
+        self.longest_prefix_to_indices = self._order_by_common_prefix(combined_ids)
+        self.indices_to_longest_prefix = OrderedDict()
+        for prefix in sorted(self.longest_prefix_to_indices.keys(), key = lambda x : -len(x)):
+            # indices for each prefix are already sorted by trie
+            for index in self.longest_prefix_to_indices[prefix]:
+                self.indices_to_longest_prefix[index] = prefix
+        return list(self.indices_to_longest_prefix.keys())
+    
+    def _order_by_common_prefix(self, sequences: Sequence[Sequence[int]]) -> Dict[Sequence[Optional[int]],Sequence[int]]:
+        longest_prefix_to_indices: Dict[Sequence[Optional[int]],Sequence[int]] = {}
+        trie = PrefixTrie(sequences)
+        leaves = trie.get_leaf_nodes()
+        leaves_sequences = [tuple(leaf.get_sequence()) for leaf in leaves]
+        leaves_and_sequences = Tqdm.tqdm(zip(leaves_sequences, leaves), desc="Finding prefixes", total=len(leaves))
+        leaves2prefixes = {leaf_sequence:leaf.get_prefix_indices() for leaf_sequence, leaf in leaves_and_sequences}
+
+        indices_already_assigned = set()
+        for leaf_sequence in sorted(leaves_sequences, key=lambda leaf_sequence : -leaves2prefixes[leaf_sequence][1]):
+            prefix_indices, _ = leaves2prefixes[leaf_sequence]
+            prefix_indices = [prefix_index for prefix_index in prefix_indices if prefix_index not in indices_already_assigned]
+            indices_already_assigned.update(prefix_indices)
+            if len(prefix_indices) > 0:
+                longest_prefix_to_indices[leaf_sequence] = tuple(prefix_indices)
+
+        return longest_prefix_to_indices
+
+    def _reorder_by_longest(self, tokenized_contexts: BatchEncoding, tokenized_continuations: BatchEncoding) -> Sequence[int]:
+        assert len(tokenized_contexts['input_ids']) == len(tokenized_continuations['input_ids'])
+        lengths = torch.tensor([
+            len(tokenized_contexts["input_ids"][i]) + len(tokenized_continuations["input_ids"][i])
+            for i in range(len(tokenized_contexts['input_ids']))
+        ], dtype=torch.int)
+        return torch.argsort(lengths, descending=True).tolist()
+
+    def _reset_cache_variables(self):
+        self.cached_sequence: Sequence[Optional[int]] = None
+        self.cached_past_key_values: torch.Tensor = None
+        self.longest_prefix_to_indices: Dict[Sequence[Optional[int]],Sequence[int]] = None
+        self.indices_to_longest_prefix: OrderedDict[int,Sequence[int]] = None
```

Comment:
As we discussed, @dirkgr I'll make these attributes into a separate cache object and pass that around thru args instead of relying on the object state of the model.

Reply:
resolved in 8fbf35994f3841ae1e5be7dbade2d92790afa2f9